src/charm.py

```diff
@@ -0,0 +1,581 @@
+        """Returns whether the pfcp agent config file was written to the workload container.
+
+        Returns:
+            bool: Whether the pfcp agent config file was written
+        """
+        if not self._pfcp_agent_container.exists(
+            f"{PFCP_AGENT_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"
+        ):
+            return False
+        return True
+
+    def _configure_bessd_workload(self, event: EventBase) -> None:
+        """Handles events for configuring bessd container.
+
+        Args:
+            event: EventBase
+        """
+        if invalid_configs := self._get_invalid_configs():
+            self.unit.status = BlockedStatus(
+                f"The following configurations are not valid: {invalid_configs}"
+            )
+            return
+        if not self._kubernetes_multus.multus_is_configured():
+            self.unit.status = WaitingStatus("Waiting for statefulset to be patched")
+            event.defer()
+            return
+        if not self._bessd_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to be able to connect to the container")
+            event.defer()
+            return
+        if not self._has_net_admin_capability():
+            self.unit.status = WaitingStatus("Waiting for statefulset to be patched")
+            event.defer()
+            return
+        if not self._bessd_container.exists(path=BESSD_CONTAINER_CONFIG_PATH):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        self._write_bessd_config_file(
+            upf_hostname=self._upf_hostname,
+            upf_mode=UPF_MODE,
+            access_interface_name=ACCESS_INTERFACE_NAME,
+            core_interface_name=CORE_INTERFACE_NAME,
+            dnn=self._dnn,  # type: ignore[arg-type]
+            pod_share_path=POD_SHARE_PATH,
+        )
+        self._create_access_route()
+        self._create_core_route()
+        if not self._ip_tables_rule_exists():
+            self._create_ip_tables_rule()
+        self._bessd_container.add_layer("upf", self._bessd_pebble_layer, combine=True)
+        self._bessd_container.replan()
+        self._bessd_container.restart(self._bessd_service_name)
+        self._run_bess_configuration()
+        self._set_unit_status()
+
+    def _run_bess_configuration(self) -> None:
+        """Runs bessd configuration in workload."""
+        initial_time = time.time()
+        timeout = 30
+        while time.time() - initial_time <= timeout:
+            try:
+                self._exec_command_in_bessd_workload(
+                    command="bessctl run /opt/bess/bessctl/conf/up4",
+                    environment=self._bessd_environment_variables,
+                )
+                return
+            except ExecError:
+                logger.info("Failed running configuration for bess")
+                time.sleep(2)
```

Comment:
```suggestion
        for _ in range(15):
            try:
                self._exec_command_in_bessd_workload(
                    command="bessctl run /opt/bess/bessctl/conf/up4",
                    environment=self._bessd_environment_variables,
                )
                return
            except ExecError:
                logger.info("Failed running configuration for bess")
                time.sleep(2)
```

This is a lot simpler. If you want something more explicit, I think it would be worth it to consider this library: https://github.com/jd/tenacity.

Reply:
I don't know about the "a lot simpler" and the behavior is different. We used to have an explicit 30 seconds timeout, now we have a 15 tries and no "time" constraint. Ex. if `_exec_command_in_bessd_workload` takes 10 seconds each time, it makes for a 180 second loop (15*(10+2)).